### ISSUE TITLE
Fix Delete Chain

### ISSUE THREAD

**Issue by alexyoung23j** [hicommonwealth/commonwealth#4167]
BD has reported failures with the delete chain route once again.

**Acceptance Criteria:**
Delete chain works as expected (on the god mode panel).

**Comment by alexyoung23j** [comment]
Okay analysis of the problem (thanks @NakulManchanda for pairing!):

We have a paranoid delete set up for Threads meaning we just update the deleted_at column when a thread is being deleted. As part of the delete chain route we obviously need to delete the threads and in another step all addresses associated with the chain. Now in the case that an address has made a thread in that community, we hit a foreign key constraint because the address_id column on thread still exists (due to paranoid deletion) but we are attempting to remove that address. 

@jnaviask @CowMuon Let me know what you think is the best fix here. We could remove the paranoid delete, remove the foreign key constraint, simply not delete the addresses, etc.

**Comment by NakulManchanda** [comment]
We just need to consistent -
- This require a cleanup
  - we have very hard Foreign Key constraints
  - In this case, "Threads" table has is logical delete setup
  - All its foreign keys tables are complaining on being dropped - in this case "Addresses"
- Root cause is logical delete can only be setup 
     - if table has foreign key table we have to work around foreign key constraint if parent needs to dropped

There can multiple solution possible:
- First, we can **physical delete "Threads"** - > then all its children for whom thread_id is foreign key can complain
- Second, make **"Addresses" logical delete** as well - then we may run into problem when deleting Chain - because chain is foreign key of "Addresses"
- Third, in "Threads" 
  - we can **relax a foreign key constraint to allow null** (need to check), and mark address_id  col in "Threads" as NULL
  - if need to keep original author address - we can create new column deleted_author
  - basically **all deleted threads will have author as NULL**
  
Order of delete need to resolved.
- foreign key table becomes parent - Address is parent of Thread
- parent rows cannot be dropped without all its children being dropped first

I have a similar issue with "Subscriptions" - https://github.com/hicommonwealth/commonwealth/issues/4033
  - user unable to delete subscription because of foreign key constraints to NotificationRead
  - in that case basically we can just add delete for "NotificationRead" as not a primary entity anyways

**Comment by alexyoung23j** [comment]
Two relevant user stories to inform the decision here:

1) God mode admin wants to _actually_ delete the community (no expectation of it returning in a new form under the same name). This is the straightforward use of deleteChain, and in this context we don't technically care if the community id can be reused later.

2) God mode admin wants to modify some data about a community or change the slug, and is using deleteChain to remove the existing community (id named `my-id`) and create a new community `my-id` with some information changed. It would be preferable to allow community to simply change all metadata without this "hack", but atm we need to work with this hack as a part of BD's workflow. In this case, we definitely do care that the community id can be reused later.

Solution here needs to enable both of these stories.

**Comment by NakulManchanda** [comment]
Good summary is:
- logically delete have cascading affect, all parent must be logically deleted too
- physical delete cascade in other direction - if parent physically delete all its children should be physically deleted

One potential work around for logical delete is to stop have this cascading 
- allow nulls for foreign key columns - all foreign key references must be removed to allow parents to delete

**Comment by alexyoung23j** [comment]
Discussed with nakul and think probably best solution given product and eng constraints is:

- Create a new column on Threads `created_by` to act as a proxy for existing column `address_id`. This field will store the string address associated with the Addresses entry with id=address_id. To accomplish this, we need a migration to fill this in for all existing Threads, and an update to the createThread route to populate `created_by`.
- Then, we can allow the foreign key `address_id` on Threads to be nullable, and add the following SQL to automatically set it null when an Address is physically deleted: `ALTER TABLE "Threads" ADD CONSTRAINT "OffchainThreads_author_id_fkey" FOREIGN KEY ("address_id") REFERENCES "Addresses"(id) ON UPDATE CASCADE ON DELETE SET NULL;`. This will prevent the foreign key constraint from causing the deleteChain route to fail when attempting to delete Addresses after deleting Threads.
